### PR TITLE
feat: add CreateSearchFilter API

### PR DESCRIPTION
**Pull request by kakcy**
This pull request introduces a new feature for managing search filters in the account service. The changes include adding new status codes, implementing the `CreateSearchFilterV2` method, adding validation for search filter creation requests, updating the command handler, and modifying the protobuf definitions.

### New Feature: Search Filters Management

* [`pkg/account/api/search_filter.go`](diffhunk://#diff-5a14057d06ba5c11b546e74297a17b313440c89848761643556a46bcdae3f7e9R1-R123): Implemented the `CreateSearchFilterV2` method for creating search filters, including necessary imports and logging.
* [`pkg/account/api/validation.go`](diffhunk://#diff-0de67f64830366bcb37ac25ac98abe4f2e03ef06f8ec8a05f78798e5b18b3957R502-R578): Added `validateCreateSearchFilterRequest` function to validate the search filter creation request.

part of #721

### CODE REVIEW

File: pkg/domainevent/domain/message.go [hunk 1]
```diff
@@ -557,6 +557,22 @@ func LocalizedMessage(eventType proto.Event_Type, localizer locale.Localizer) *p
...
+			Message: localizer.MustLocalizeWithTemplate(
+				locale.CreatedTemplate,
+				localizer.MustLocalizeWithTemplate(locale.AccountSearchFilter),
+			),
+		}
+	case proto.Event_ACCOUNT_V2_UPDATED_SEARCH_FILTER:
+		return &proto.LocalizedMessage{
+			Locale: localizer.GetLocale(),
+			Message: localizer.MustLocalizeWithTemplate(
+				locale.UploadedTemplate,
```
Comment (cre8ivejp) [inline pkg/domainevent/domain/message.go:572]:
```suggestion
				locale.UppdatedTemplate,
```

File: proto/event/domain/event.proto [hunk 2]
```diff
@@ -596,6 +598,16 @@ message AccountV2DeletedEvent {
   string email = 1;
 }
 
+message SearchFilterCreateEvent {
```
Comment (cre8ivejp) [inline proto/event/domain/event.proto:601]: The query, type, and other fields are missing.

File: proto/event/domain/event.proto [hunk 3]
```diff
@@ -596,6 +598,16 @@ message AccountV2DeletedEvent {
...
+  string id = 1;
+  string Name = 2;
+}
+
+message SearchFilterUpdateEvent {
```
Comment (cre8ivejp) [inline proto/event/domain/event.proto:606]: Same here.

File: proto/event/domain/event.proto (hunk 2 above)
Comment (cre8ivejp) [inline proto/event/domain/event.proto:601]:
```suggestion
message SearchFilterCreatedEvent {
```

File: proto/event/domain/event.proto (hunk 3 above)
Comment (cre8ivejp) [inline proto/event/domain/event.proto:606]:
```suggestion
message SearchFilterUpdatedEvent {
```

File: pkg/account/api/search_filter.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+		)
+		return nil, err
+	}
+
+	account, err := s.getAccountV2(ctx, req.Email, req.OrganizationId, localizer)
+	if err != nil {
+		return nil, err
+	}
+	// Since there is only one default setting for a filter target, set the existing default to OFF.
+	changeDefaultFilters := getChangeDefaultFilters(account, req.Command.SearchFilter)
```
Comment (cre8ivejp) [inline pkg/account/api/search_filter.go:61]: It's better not to change the default setting for existing filters. When creating a new filter, we don't necessarily want to be the default setting.

Comment (kakcy) [reply pkg/account/api/search_filter.go:61]: @cre8ivejp 

> When creating a new filter, we don't necessarily want to be the default setting.

As you said, I know there is no need to set default settings at creation time.

The process here is to avoid having multiple default filters.

So, if you already have a filter with default settings and you want to set the newly created filter as the default, I think you need to turn off the previous default settings. What do you think?

**Comment by kakcy** [comment]
I have fixed the points raised.
Also, I have a comment for your suggestion, so please check them when you are free.

File: pkg/account/command/account_v2.go [hunk 5]
```diff
@@ -182,6 +186,37 @@ func (h *accountV2CommandHandler) delete(ctx context.Context, _ *accountproto.De
 	})
 }
 
+func (h *accountV2CommandHandler) createSearchFiler(
```
Comment (cre8ivejp) [inline pkg/account/command/account_v2.go:189]:
```suggestion
func (h *accountV2CommandHandler) createSearchFilter(
```

File: pkg/account/command/account_v2.go [hunk 6]
```diff
@@ -182,6 +186,37 @@ func (h *accountV2CommandHandler) delete(ctx context.Context, _ *accountproto.De
...
+		cmd.Name,
+		cmd.Query,
+		cmd.FilterTargetType,
+		cmd.EnvironmentId,
+		cmd.DefaultFilter); err != nil {
+		return err
+	}
+	return h.send(ctx, eventproto.Event_ACCOUNT_V2_CREATED_SEARCH_FILTER, &eventproto.SearchFilterCreatedEvent{
+		Id:   searchFilter.Id,
+		Name: searchFilter.Name,
```
Comment (cre8ivejp) [inline pkg/account/command/account_v2.go:204]: The query and other fields are missing.

File: pkg/account/command/account_v2.go [hunk 7]
```diff
@@ -182,6 +186,37 @@ func (h *accountV2CommandHandler) delete(ctx context.Context, _ *accountproto.De
...
+
+func (h *accountV2CommandHandler) updateSearchFiler(
+	ctx context.Context,
+	cmd *accountproto.UpdateSearchFilterCommand) error {
+	if err := h.account.UpdateSearchFilter(cmd.SearchFilter); err != nil {
+		return err
+	}
+	return h.send(ctx, eventproto.Event_ACCOUNT_V2_CREATED_SEARCH_FILTER, &eventproto.SearchFilterUpdatedEvent{
+		Id:   cmd.SearchFilter.Id,
+		Name: cmd.SearchFilter.Name,
```
Comment (cre8ivejp) [inline pkg/account/command/account_v2.go:216]: Same here.

File: proto/account/command.proto [hunk 8]
```diff
@@ -68,3 +69,15 @@ message ChangeAPIKeyNameCommand {
 message EnableAPIKeyCommand {}
 
 message DisableAPIKeyCommand {}
+
+message CreateSearchFilterCommand {
+  string name = 1;
+  string query = 2;
+  account.FilterTargetType filter_target_type = 3;
+  string environment_id = 4;
```
Comment (cre8ivejp) [inline proto/account/command.proto:77]: We don't need the environment ID in the command. It's set in the Request.

File: proto/account/command.proto [hunk 9]
```diff
@@ -68,3 +69,15 @@ message ChangeAPIKeyNameCommand {
...
+  bool default_filter = 5;
+}
+
+message UpdateSearchFilterCommand {
+  SearchFilter search_filter = 1;
```
Comment (cre8ivejp) [inline proto/account/command.proto:82]: As discussed, we will only pass the fields which will be updated like in the `CreateSearchFilterCommand`.

File: proto/account/service.proto [hunk 10]
```diff
@@ -215,721 +213,42 @@ message GetAPIKeyBySearchingAllEnvironmentsResponse {
...
+  string email = 1;
+  string organization_id = 2;
+  string environment_namespace = 3;
+  CreateSearchFilterCommand command = 4;
+}
+
+message CreateSearchFilterResponse {}
 
 service AccountService {
-  rpc GetMe(GetMeRequest) returns (GetMeResponse) {
```
Comment (cre8ivejp) [inline proto/account/service.proto:234]: Why did you remove the API descriptions?

File: proto/account/service.proto [hunk 11]
```diff
@@ -215,721 +213,42 @@ message GetAPIKeyBySearchingAllEnvironmentsResponse {
...
-        key: "application/json"
-        value: '{ "code": 13, "message": "internal", "details": [] }'
-      }
-    }
-  }
-};
+message CreateSearchFilterRequest {
+  string email = 1;
+  string organization_id = 2;
+  string environment_namespace = 3;
```
Comment (cre8ivejp) [inline proto/account/service.proto:219]: For new APIs we use `environment_id`.

```suggestion
  string environment_id = 3;
```

File: proto/account/command.proto (hunk 8 above)
Comment (kakcy) [reply proto/account/command.proto:77]: I understand that filters are saved separately for each Env, but is this correct?
In that case, it is necessary to retain Env information, so I think it is necessary to include Env information in the command.What do you think?
Or is account information saved as a record for each Env?

File: proto/account/service.proto (hunk 10 above)
Comment (kakcy) [reply proto/account/service.proto:234]: I thought the code had changed when I ran the make command, so I committed it. I'll change it back.

File: proto/account/command.proto (hunk 8 above)
Comment (cre8ivejp) [reply proto/account/command.proto:77]: Sorry. It's correct as it is.

File: pkg/account/api/search_filter.go [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+	}
+
+	account, err := s.getAccountV2(ctx, req.Email, req.OrganizationId, localizer)
+	if err != nil {
+		return nil, err
+	}
+	// Since there is only one default setting for a filter target, set the existing default to OFF.
+	commands := make([]command.Command, 0)
+	if req.Command.DefaultFilter {
+		changeDefaultFilterCmds := getChangeDefaultSearchFilterCommands(
```
Comment (kakcy) [inline pkg/account/api/search_filter.go:63]: I created a new `ChangeDefaultSearchFilterCommand` and changed it to specify the ID and update only the `DefaultFilter`.

**Review by kakcy — COMMENTED** [review]
I added docs for WebAPI.
https://github.com/bucketeer-io/bucketeer/pull/1210/commits/fde8e809390fd6a46da118a1300283004c47c1e9

File: proto/account/service.proto [hunk 13]
```diff
@@ -932,4 +941,50 @@ service AccountService {
...
   }
+  rpc CreateSearchFilter(CreateSearchFilterRequest)
+      returns (CreateSearchFilterResponse) {
+    option (google.api.http) = {
+      post: "/v1/account/create_search_filter"
+      body: "*"
+    };
+    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
+      summary: "Create Search Filter",
+      description: "Create a search filter to use in the console.",
```
Comment (cre8ivejp) [inline proto/account/service.proto:952]:
```suggestion
      description: "Create a search filter. To call this API, you need an `EDITOR` role.",
```

Comment (cre8ivejp) [reply proto/account/service.proto:952]: Usually, a viewer can only read data on the console. They are not allowed to create or update anything.
But since this is the user's settings, it should be fine to use Viewer for this case.

File: pkg/account/api/search_filter.go [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+	for _, filter := range account.SearchFilters {
+		if filter.DefaultFilter &&
+			filterTarget == filter.FilterTargetType &&
+			environmentId == filter.EnvironmentId {
+			changeDefaultFiltersCmds = append(changeDefaultFiltersCmds, &accountproto.ChangeDefaultSearchFilterCommand{
+				Id:            filter.Id,
+				DefaultFilter: false,
+			})
+		}
+	}
```
Comment (cre8ivejp) [inline pkg/account/api/search_filter.go:121]: The API layer only validates the request and processes the commands.
This logic should be in the domain layer when adding and updating a filter.

Also, the `ChangeDefaultSearchFilterCommand` should be used only for updating, not for creating.

**Comment by kakcy** [comment]
> The API layer only validates the request and processes the commands.
> This logic should be in the domain layer when adding and updating a filter.

I Fixed the turning OFF of DefaultFilter to be done at the domain layer.
https://github.com/bucketeer-io/bucketeer/pull/1210/commits/3f97f81080e57dabbc91626383533b898d6e1818

File: pkg/account/domain/account.go [hunk 15]
```diff
@@ -118,23 +118,27 @@ func (a *AccountV2) AddSearchFilter(
...
+		a.ternOffDefaultFilters(targetType, environmentId)
 	}
 
 	searchFilter := &proto.SearchFilter{
 		Id:               id.String(),
 		Name:             name,
 		Query:            query,
 		FilterTargetType: targetType,
-		EnvironmentId:    environmentID,
+		EnvironmentId:    environmentId,
```
Comment (cre8ivejp) [inline pkg/account/domain/account.go:136]: Same here.

File: pkg/account/api/search_filter.go [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+			}
+			return nil, dt.Err()
+		}
+		s.logger.Error(
+			"Failed to create search filter",
+			log.FieldsFromImcomingContext(ctx).AddFields(
+				zap.Error(err),
+				zap.String("organizationID", req.OrganizationId),
+				zap.String("email", req.Email),
+				zap.String("searchFilterName", req.Command.Name),
```
Comment (cre8ivejp) [inline pkg/account/api/search_filter.go:76]: Please also add the query and search type in the logs.

File: pkg/account/command/account_v2.go [hunk 17]
```diff
@@ -72,6 +73,12 @@ func (h *accountV2CommandHandler) Handle(ctx context.Context, cmd Command) error
 		return h.disable(ctx, c)
 	case *accountproto.DeleteAccountV2Command:
 		return h.delete(ctx, c)
+	case *accountproto.CreateSearchFilterCommand:
+		fmt.Printf("account CreateSearchFilterCommand\n")
+		return h.createSearchFilter(ctx, c)
+	case *accountproto.ChangeDefaultSearchFilterCommand:
+		fmt.Printf("account ChangeDefaultSearchFilterCommand\n")
```
Comment (cre8ivejp) [inline pkg/account/command/account_v2.go:80]:
```suggestion
```

File: pkg/account/command/account_v2.go [hunk 18]
```diff
@@ -72,6 +73,12 @@ func (h *accountV2CommandHandler) Handle(ctx context.Context, cmd Command) error
 		return h.disable(ctx, c)
 	case *accountproto.DeleteAccountV2Command:
 		return h.delete(ctx, c)
+	case *accountproto.CreateSearchFilterCommand:
+		fmt.Printf("account CreateSearchFilterCommand\n")
```
Comment (cre8ivejp) [inline pkg/account/command/account_v2.go:77]:
```suggestion
```

File: pkg/account/command/account_v2.go [hunk 19]
```diff
@@ -182,6 +189,43 @@ func (h *accountV2CommandHandler) delete(ctx context.Context, _ *accountproto.De
...
+	return h.send(ctx, eventproto.Event_ACCOUNT_V2_CREATED_SEARCH_FILTER, &eventproto.SearchFilterCreatedEvent{
+		Name:          searchFilter.Name,
+		Query:         searchFilter.Query,
+		TargetType:    searchFilter.FilterTargetType,
+		EnvironmentId: searchFilter.EnvironmentId,
+		DefaultFilter: searchFilter.DefaultFilter,
+	})
+}
+
+func (h *accountV2CommandHandler) updateDefaultSearchFilter(
```
Comment (cre8ivejp) [inline pkg/account/command/account_v2.go:213]:
```suggestion
func (h *accountV2CommandHandler) changeDefaultSearchFilter(
```

File: pkg/account/command/account_v2.go [hunk 20]
```diff
@@ -182,6 +189,43 @@ func (h *accountV2CommandHandler) delete(ctx context.Context, _ *accountproto.De
...
+	ctx context.Context,
+	cmd *accountproto.ChangeDefaultSearchFilterCommand) error {
+	if err := h.account.UpdateDefaultSearchFilter(cmd.Id, cmd.DefaultFilter); err != nil {
```
Comment (cre8ivejp) [inline pkg/account/command/account_v2.go:216]:
```suggestion
	if err := h.account.ChangeDefaultSearchFilter(cmd.Id, cmd.DefaultFilter); err != nil {
```

File: pkg/account/domain/account.go [hunk 21]
```diff
@@ -118,23 +118,27 @@ func (a *AccountV2) AddSearchFilter(
 	name string,
 	query string,
 	targetType proto.FilterTargetType,
-	environmentID string, defaultFilter bool) error {
+	environmentId string, defaultFilter bool) (*proto.SearchFilter, error) {
```
Comment (cre8ivejp) [inline pkg/account/domain/account.go:121]: In Go, the id is in upper case.

```suggestion
	environmentID string, defaultFilter bool) (*proto.SearchFilter, error) {
```

File: pkg/account/domain/account.go [hunk 22]
```diff
@@ -151,13 +155,55 @@ func (a *AccountV2) DeleteSearchFilter(id string) error {
 	return errSearchFilterNotFound
 }
 
-func (a *AccountV2) UpdateSearchFilter(searchFilter *proto.SearchFilter) error {
-	for i, f := range a.AccountV2.SearchFilters {
-		if f.Id == searchFilter.Id {
-			a.AccountV2.SearchFilters[i] = searchFilter
+func (a *AccountV2) UpdateDefaultSearchFilter(id string, defaultFilter bool) error {
```
Comment (cre8ivejp) [inline pkg/account/domain/account.go:158]:
```suggestion
func (a *AccountV2) ChangeDefaultSearchFilter(id string, defaultFilter bool) error {
```

File: pkg/account/domain/account.go [hunk 23]
```diff
@@ -151,13 +155,55 @@ func (a *AccountV2) DeleteSearchFilter(id string) error {
...
+				DefaultFilter:    defaultFilter,
+			}
 			a.UpdatedAt = time.Now().Unix()
 			return nil
 		}
 	}
 	return errSearchFilterNotFound
 }
+
+func (a *AccountV2) ternOffDefaultFilters(targetFilter proto.FilterTargetType, environmentId string) {
```
Comment (cre8ivejp) [inline pkg/account/domain/account.go:201]:
```suggestion
func (a *AccountV2) resetDefaultFilter(targetFilter proto.FilterTargetType, environmentID string) {
```

File: pkg/account/domain/account.go [hunk 24]
```diff
@@ -151,13 +155,55 @@ func (a *AccountV2) DeleteSearchFilter(id string) error {
...
+
+			f.DefaultFilter = defaultFilter
+			a.UpdatedAt = time.Now().Unix()
+			return nil
+		}
+	}
+	return errSearchFilterNotFound
+}
+
+func (a *AccountV2) UpdateSearchFilter(
```
Comment (cre8ivejp) [inline pkg/account/domain/account.go:174]: Do we need this function?
We have a function to change the default filter.
If you plan to update the name and query for the existing filters, we should use `changeName` and `changeQuery` instead.

**Review by cre8ivejp — APPROVED** [review]
Thank you 🎉
